## Predicción del Precio de Casas en Boston - Entendimiento del Problema y Obtención de Datos

### By:
RRafael Garcia

### Date:
2026-08-15

### Description:
Requerimiento
Crear una nueva rama de git (Usar Gitflow) y Crear un notebook para la obtención de datos tipo RAW.

crear la rama a partir de un issue: https://docs.github.com/es/issues/tracking-your-work-with-issues/using-issues/creating-a-branch-for-an-issue

Realizar los pasos de: https://joserzapata.github.io/post/ciencia-datos-proyecto-python/1-data/

Descargar los datos y entender el problema a realizar, contestar en el notebook

¿Cual es el objetivo del problema?
¿Cómo se usará su solución?
¿Cuáles son las soluciones actuales (si las hay)?
¿Cómo se debe enmarcar este problema (supervisado / no supervisado, en línea / fuera de línea, etc.)
¿Cómo se debe medir el desempeño o el rendimiento de la solución, una primera intuicion?
¿La medida de desempeño está alineada con el objetivo del problema?
¿Cuál sería el desempeño o rendimiento mínimo necesario para alcanzar el objetivo del problema?
¿Cuáles son los problemas parecidos? ¿Se puede reutilizar experiencias o herramientas ya creadas?
¿Hay experiencia del problema disponible?
(Importante) ¿Cómo se puede resolver el problema manualmente?
Hacer un listado de los supuestos que hay hasta este momento.
Cual es la fuente de los datos?
Como se actualizan los datos?
Cada cuanto tiempo se actualizan los datos
Entregables
Notebook para obtención de los datos.

Se debe realizar un Pull request para ingresar el notebook a la ramamain para esto debe tener mínimo 1 revisiones de otras personas del Curso y que pase los check del CI/CD

## 📚 Import libraries

In [ ]:
import pandas as pd
import numpy as np
import os
from pathlib import Path


/home/rafaelgarcia11/Documentos/Universidad_Maestria/MLOPS/Boston_Home_Prices/notebooks/1-data


## 💾 Load data

In [ ]:
# Cargar el dataset
df = pd.read_csv("../../data/01_raw/Precios_Casas_Boston.csv")
print(f"Dimensiones del dataset: {df.shape[0]} filas x {df.shape[1]} columnas")
df.head(10)

Dimensiones del dataset: 448 filas x 15 columnas


,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
0,1,0.00632,18.0,2.31,0.0,538.0,6575.0,65.2,4.0900,1.0,296.0,15.3,396.90,4.98,24.0
1,2,0.02731,0.0,7.07,0.0,469.0,6421.0,78.9,4.9671,2.0,242.0,17.8,396.90,9.14,21.6
2,4,0.03237,0.0,2.18,0.0,458.0,6998.0,45.8,6.0622,3.0,222.0,18.7,394.63,2.94,33.4
3,5,0.06905,0.0,2.18,0.0,458.0,7147.0,54.2,6.0622,3.0,222.0,18.7,396.90,5.33,NaN
4,7,0.08829,12.5,7.87,0.0,524.0,6012.0,66.6,5.5605,5.0,311.0,15.2,395.60,12.43,22.9
5,11,0.22489,12.5,7.87,0.0,524.0,6377.0,94.3,6.3467,5.0,311.0,15.2,392.52,20.45,15.0
6,12,0.11747,12.5,7.87,0.0,524.0,6009.0,82.9,6.2267,5.0,311.0,15.2,396.90,13.27,NaN
7,13,0.09378,12.5,7.87,0.0,524.0,5889.0,39.0,5.4509,5.0,311.0,15.2,390.50,15.71,21.7
8,14,0.62976,0.0,8.14,0.0,538.0,5949.0,61.8,4.7075,4.0,307.0,21.0,396.90,8.26,20.4
9,15,0.63796,0.0,8.14,0.0,538.0,6096.0,84.5,4.4619,4.0,307.0,21.0,380.02,10.26,18.2


In [14]:
# Revisión general de los tipos de datos y valores nulos
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 448 entries, 0 to 447
Data columns (total 15 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   ID       448 non-null    int64  
 1   crim     443 non-null    float64
 2   zn       446 non-null    float64
 3   indus    442 non-null    float64
 4   chas     446 non-null    float64
 5   nox      444 non-null    float64
 6   rm       442 non-null    float64
 7   age      444 non-null    float64
 8   dis      442 non-null    float64
 9   rad      444 non-null    float64
 10  tax      443 non-null    float64
 11  ptratio  446 non-null    float64
 12  black    445 non-null    float64
 13  lstat    447 non-null    float64
 14  medv     432 non-null    float64
dtypes: float64(14), int64(1)
memory usage: 52.6 KB


In [15]:
# Estadísticas descriptivas
df.describe().round(2)

,ID,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,black,lstat,medv
count,448.00,443.00,446.00,442.00,446.00,444.00,442.00,444.00,442.00,444.0,443.00,446.00,445.00,447.00,432.00
mean,263.93,75.81,9.15,12.03,0.06,479.45,5590.95,71.26,352.52,9.8,418.36,18.40,347.80,13.19,21.99
std,150.88,894.07,20.79,7.07,0.23,227.80,1991.24,27.67,1165.06,8.8,171.61,2.24,99.92,7.24,9.22
min,1.00,0.01,0.00,0.74,0.00,0.40,4.97,6.00,1.13,1.0,188.00,12.60,3.50,1.73,5.00
25%,140.00,0.09,0.00,5.86,0.00,433.00,5710.00,50.68,2.02,4.0,284.50,16.80,368.74,7.52,16.10
50%,265.50,0.35,0.00,10.01,0.00,532.00,6098.50,82.15,3.10,5.0,384.00,19.00,391.23,12.03,20.60
75%,410.25,4.55,0.00,18.10,0.00,614.00,6437.75,95.00,5.50,24.0,666.00,20.20,395.99,17.28,24.70
max,516.00,15288.00,100.00,27.74,1.00,871.00,8725.00,100.00,8344.00,24.0,711.00,21.20,396.90,37.97,50.00


## 🔍 Entendimiento del Problema

A continuación se responden las preguntas requeridas para comprender el problema antes de iniciar el proceso de ciencia de datos.

### 1. ¿Cuál es el objetivo del problema?

El objetivo es **predecir el valor mediano de las viviendas ocupadas por sus propietarios** (`medv`, en miles de dólares) en distintas zonas de Boston, a partir de 13 atributos que describen características socioeconómicas y geográficas de cada zona (tasa de criminalidad, número promedio de habitaciones, distancia a centros de empleo, proporción de terreno residencial, entre otros).

Se trata de un problema de **regresión** donde la variable objetivo es continua.

### 2. ¿Cómo se usará su solución?

La solución puede utilizarse como una herramienta de estimación de precios de viviendas. Potenciales usos incluyen:

- **Agentes inmobiliarios**: estimar el precio de mercado de una vivienda en función de las características de la zona.
- **Compradores/vendedores**: obtener una referencia de precio antes de una transacción.
- **Entidades financieras**: evaluar el valor de una propiedad como garantía para créditos hipotecarios.
- **Urbanistas y gobierno local**: analizar qué factores inciden más en el valor de las viviendas para orientar políticas públicas.

En el contexto de este curso, la solución sirve como **prueba de concepto (POC)** para aplicar un proceso completo de ciencia de datos con enfoque MLOps.

### 3. ¿Cuáles son las soluciones actuales (si las hay)?

- **Avalúos manuales**: peritos inmobiliarios que visitan la propiedad y la valoran con base en su experiencia, comparables del mercado y tablas de referencia.
- **Modelos hedónicos de precios**: modelos de regresión lineal que descomponen el precio en función de los atributos de la vivienda y su entorno. Son ampliamente usados en economía inmobiliaria.
- **Plataformas inmobiliarias** (Zillow Zestimate, Redfin, etc.): utilizan modelos de machine learning a gran escala con datos actualizados del mercado.
- **Soluciones académicas sobre este mismo dataset**: el dataset de Boston Housing es uno de los más clásicos en ML. Se han publicado múltiples soluciones con regresión lineal, árboles de decisión, Random Forest, Gradient Boosting y redes neuronales.

### 4. ¿Cómo se debe enmarcar este problema (supervisado/no supervisado, en línea/fuera de línea, etc.)?

- **Tipo de aprendizaje**: Supervisado, ya que se dispone de una variable objetivo conocida (`medv`).
- **Tipo de tarea**: Regresión, puesto que la variable a predecir es continua (precio en miles de dólares).
- **Modalidad**: Fuera de línea (batch), ya que el dataset es estático y no se requiere actualización en tiempo real.
- **Salida**: Univariada — se predice una única variable numérica.

### 5. ¿Cómo se debe medir el desempeño o el rendimiento de la solución, una primera intuición?

Para un problema de regresión, las métricas candidatas son:

| Métrica | Descripción | Ventaja |
|---------|-------------|--------|
| **RMSE** (Root Mean Squared Error) | Raíz del error cuadrático medio | Penaliza errores grandes, mismas unidades que la variable objetivo (miles USD) |
| **MAE** (Mean Absolute Error) | Error absoluto medio | Más robusta a outliers, fácil de interpretar |
| **R²** (Coeficiente de determinación) | Proporción de varianza explicada | Valor entre 0 y 1, fácil de comunicar |
| **MAPE** (Mean Absolute Percentage Error) | Error porcentual absoluto medio | Interpretación relativa al precio |

**Primera intuición**: usar **RMSE** como métrica principal ya que penaliza los errores grandes de predicción (que en un contexto inmobiliario son los más costosos) y se expresa en las mismas unidades que el precio (miles de dólares). Complementar con **R²** para comunicar el rendimiento del modelo de forma intuitiva.

### 6. ¿La medida de desempeño está alineada con el objetivo del problema?

Sí. El RMSE captura directamente el error promedio en la predicción del precio y penaliza las desviaciones grandes, lo cual es coherente con el objetivo de estimar precios de viviendas de forma confiable. Un error grande en la estimación de un precio puede tener consecuencias financieras significativas, por lo que una métrica que penalice estos errores resulta adecuada.

El R² complementa indicando qué proporción de la variabilidad del precio es capturada por el modelo, lo cual es útil para evaluar si el modelo aporta valor frente a simplemente usar el promedio.

### 7. ¿Cuál sería el desempeño o rendimiento mínimo necesario para alcanzar el objetivo del problema?

Considerando que la desviación estándar del precio (`medv`) es aproximadamente **$9,220 USD**, un modelo útil debería tener:

- **RMSE < $5,000 USD** (menor que la mitad de la desviación estándar del target), lo que indicaría que el modelo predice mejor que simplemente estimar con la media.
- **R² > 0.70**, lo que significaría que el modelo explica al menos el 70% de la variabilidad de los precios.

Estos umbrales se definen como el mínimo aceptable para la POC. En un contexto de producción real, se esperaría un RMSE más bajo y un R² cercano a 0.85-0.90.

### 8. ¿Cuáles son los problemas parecidos? ¿Se puede reutilizar experiencias o herramientas ya creadas?

**Problemas similares:**
- Predicción de precios de viviendas en Kaggle (House Prices - Advanced Regression Techniques, California Housing).
- Estimación de precios de vehículos usados, seguros, o alquiler de propiedades.
- Cualquier problema de regresión con variables mixtas (numéricas y categóricas).

**Experiencias y herramientas reutilizables:**
- Pipelines de `scikit-learn` para preprocesamiento y modelado.
- Técnicas de feature engineering estándar para datos tabulares (escalado, encoding, imputación).
- Modelos probados en problemas similares: Random Forest, Gradient Boosting (XGBoost, LightGBM), regresión regularizada (Lasso, Ridge, ElasticNet).
- Librerías de AutoML como PyCaret o AutoGluon.

### 9. ¿Hay experiencia del problema disponible?

Sí, este es uno de los datasets más utilizados en la historia del machine learning:

- Publicado originalmente por Harrison y Rubinfeld en 1978 en el artículo *"Hedonic housing prices and the demand for clean air"*.
- Ha sido incluido en `scikit-learn`, el repositorio UCI Machine Learning y múltiples libros de texto.
- Existe amplia documentación sobre qué modelos funcionan bien, qué atributos son más relevantes (generalmente `rm`, `lstat` y `ptratio` son los más predictivos) y qué errores típicos se obtienen.

**Nota**: este dataset ha sido objeto de debate ético por la variable `black`, que codifica composición racial. Algunas versiones recientes lo han retirado de `scikit-learn` por esta razón.

### 10. (Importante) ¿Cómo se puede resolver el problema manualmente?

Un experto inmobiliario podría estimar manualmente el precio de una vivienda en Boston usando reglas heurísticas como:

1. **Partir de un precio base** cercano a la mediana (~$20,600 USD).
2. **Ajustar según el número de habitaciones** (`rm`): a mayor número de habitaciones, mayor precio. Cada habitación adicional podría representar un incremento de ~$3,000-$5,000.
3. **Penalizar por criminalidad** (`crim`): zonas con alta tasa de criminalidad tendrían un descuento.
4. **Penalizar por proporción de población de estatus bajo** (`lstat`): a mayor porcentaje, menor precio.
5. **Considerar cercanía a centros de empleo** (`dis`): la distancia a centros laborales influye en el precio.
6. **Factor río Charles** (`chas`): las viviendas colindantes con el río tienen un precio premium.

Este enfoque manual tendría errores significativos pero serviría como **modelo heurístico base** para comparar con los modelos de ML.

### 11. Listado de supuestos hasta este momento

1. Los datos representan una muestra razonablemente representativa de las zonas de Boston.
2. La relación entre los atributos y el precio de la vivienda se puede capturar con modelos de regresión.
3. Los 13 atributos proporcionados contienen información suficiente para hacer predicciones útiles.
4. Los valores nulos encontrados pueden ser tratados sin perder información relevante (imputación o eliminación).
5. Existen posibles problemas de calidad en los datos: se observan valores con escalas inconsistentes (por ejemplo, `rm` tiene valores como 6575 en lugar de 6.575, sugiriendo multiplicación por 1000 en algunos registros).
6. El dataset es estático y no se actualiza, por lo que no hay concept drift a considerar en esta POC.
7. No hay multicolinealidad severa que impida el uso de modelos lineales (a verificar en el EDA).
8. Los IDs no consecutivos (448 registros de un rango de 1 a 516) sugieren que se eliminaron filas del dataset original, lo que podría introducir sesgo de selección.

### 12. ¿Cuál es la fuente de los datos?

El dataset original proviene del repositorio de datos del U.S. Census Service y fue recopilado por Harrison y Rubinfeld (1978). La versión utilizada en este proyecto es un archivo CSV (`Precios_Casas_Boston.csv`) proporcionado como parte del material del curso. Es una versión modificada del dataset clásico que incluye problemas de calidad de datos para fines didácticos.

### 13. ¿Cómo se actualizan los datos?

Los datos **no se actualizan**. Es un dataset histórico y estático correspondiente a información censal de la década de 1970 en Boston. Para esta POC se trabaja con el archivo CSV proporcionado sin modificaciones externas.

### 14. ¿Cada cuánto tiempo se actualizan los datos?

No aplica. El dataset es estático y no tiene frecuencia de actualización. Esto lo convierte en un caso ideal para una POC de MLOps con proceso manual, ya que no hay necesidad de gestionar pipelines de ingestión periódica.

## 👷 Data Preparation or Feature Engineering

En esta etapa (Tarea 1) no se realiza feature engineering. Sin embargo, se documenta una inspección inicial de los datos RAW para identificar problemas que deberán abordarse en las tareas posteriores.

In [16]:
# Revisión de valores nulos por columna
print("=== Valores nulos por columna ===")
null_counts = df.isnull().sum()
null_pct = (df.isnull().sum() / len(df) * 100).round(2)
null_summary = pd.DataFrame({"Nulos": null_counts, "% Nulos": null_pct})
print(null_summary[null_summary["Nulos"] > 0])

=== Valores nulos por columna ===
         Nulos  % Nulos
crim         5     1.12
zn           2     0.45
indus        6     1.34
chas         2     0.45
nox          4     0.89
rm           6     1.34
age          4     0.89
dis          6     1.34
rad          4     0.89
tax          5     1.12
ptratio      2     0.45
black        3     0.67
lstat        1     0.22
medv        16     3.57


In [20]:
# Verificar inconsistencias de escala
# Según el dataset original de Boston, 'rm' debería estar entre ~3-9
# pero se observan valores como 6575 (posiblemente rm * 1000)

SCALE_THRESHOLD = 100
RM_UPPER_LIMIT = 10

print("=== Detección de inconsistencias de escala ===")
print("\nrm (esperado ~3-9 habitaciones):")
print(f"  Valores > {SCALE_THRESHOLD}: {(df['rm'] > SCALE_THRESHOLD).sum()} registros")
print(f"  Valores < {RM_UPPER_LIMIT}:  {(df['rm'] < RM_UPPER_LIMIT).sum()} registros")

print("\nnox (esperado ~0.3-0.9 partes por 10M):")
print(f"  Valores > 1:   {(df['nox'] > 1).sum()} registros")
print(f"  Valores <= 1:  {(df['nox'] <= 1).sum()} registros")

print("\ndis (esperado ~1-13):")
print(f"  Valores > {SCALE_THRESHOLD}: {(df['dis'] > SCALE_THRESHOLD).sum()} registros")
print(f"  Valores < {SCALE_THRESHOLD}: {(df['dis'] < SCALE_THRESHOLD).sum()} registros")


=== Detección de inconsistencias de escala ===

rm (esperado ~3-9 habitaciones):
  Valores > 100: 397 registros
  Valores < 10:  45 registros

nox (esperado ~0.3-0.9 partes por 10M):
  Valores > 1:   377 registros
  Valores <= 1:  67 registros

dis (esperado ~1-13):
  Valores > 100: 44 registros
  Valores < 100: 398 registros


In [ ]:
# Verificar IDs - ¿hay registros faltantes del dataset original?
print("=== Análisis de IDs ===")
print(f"Registros presentes: {len(df)}")
print(f"Rango de ID: {df['ID'].min()} - {df['ID'].max()}")
ids_faltantes = set(range(1, df["ID"].max() + 1)) - set(df["ID"])
print(f"IDs faltantes: {len(ids_faltantes)}")
print(
    f"Esto sugiere que se eliminaron {len(ids_faltantes)} registros del dataset original."
)

=== Análisis de IDs ===
Registros presentes: 448
Rango de ID: 1 - 516
IDs faltantes: 173
Esto sugiere que se eliminaron 173 registros del dataset original.


In [ ]:
# Resumen de problemas de calidad detectados
print("=" * 60)
print("RESUMEN DE PROBLEMAS DE CALIDAD DETECTADOS")
print("=" * 60)
print()
print("1. VALORES NULOS:")
print(
    f"   - Variable objetivo (medv): {df['medv'].isnull().sum()} nulos ({df['medv'].isnull().sum() / len(df) * 100:.1f}%)"
)
print(f"   - Total de celdas con nulos: {df.isnull().sum().sum()}")
print()
print("2. INCONSISTENCIAS DE ESCALA:")
print("   - Algunas columnas tienen valores aparentemente multiplicados")
print("     por 1000 (rm, nox, dis, entre otras).")
print("   - Ejemplo: rm=6575 debería ser 6.575 habitaciones.")
print()
print("3. REGISTROS FALTANTES:")
print(f"   - {len(ids_faltantes)} IDs ausentes del rango original.")
print("   - Posible sesgo de selección a evaluar.")
print()
print("4. POSIBLES OUTLIERS:")
print(f"   - crim max: {df['crim'].max()} (media: {df['crim'].mean():.2f})")
print("   - Requiere análisis detallado en el EDA.")

RESUMEN DE PROBLEMAS DE CALIDAD DETECTADOS

1. VALORES NULOS:
   - Variable objetivo (medv): 16 nulos (3.6%)
   - Total de celdas con nulos: 66

2. INCONSISTENCIAS DE ESCALA:
   - Algunas columnas tienen valores aparentemente multiplicados
     por 1000 (rm, nox, dis, entre otras).
   - Ejemplo: rm=6575 debería ser 6.575 habitaciones.

3. REGISTROS FALTANTES:
   - 173 IDs ausentes del rango original.
   - Posible sesgo de selección a evaluar.

4. POSIBLES OUTLIERS:
   - crim max: 15288.0 (media: 75.81)
   - Requiere análisis detallado en el EDA.


## 📊 Analysis of Results and Conclusions

### Resultados de la exploración inicial

La inspección de los datos RAW revela un dataset de **448 registros y 15 columnas** (1 ID, 13 atributos predictores y 1 variable objetivo). Se identificaron los siguientes hallazgos principales:

1. **Datos faltantes**: hay valores nulos distribuidos en todas las columnas (entre 1 y 16 por columna). La variable objetivo `medv` tiene 16 valores nulos (3.6%), lo cual es significativo y requerirá una estrategia de manejo.

2. **Inconsistencias de escala**: se detectó que varias columnas (`rm`, `nox`, `dis`, entre otras) presentan valores en dos escalas distintas: la mayoría de registros parecen tener valores multiplicados por 1000 respecto a los valores esperados según la documentación del dataset original. Este problema deberá corregirse en la etapa de preparación de datos.

3. **Registros eliminados**: el dataset tiene 448 filas con IDs que van de 1 a 516, con 68 IDs ausentes. Esto indica que se eliminaron registros del dataset original, lo que podría introducir sesgo.

4. **Posibles outliers extremos**: la variable `crim` tiene un valor máximo de 15,288 frente a una media de 75.81, lo que sugiere valores atípicos muy pronunciados.

### Conclusiones

- El problema está bien definido como **regresión supervisada** con variable objetivo `medv`.
- El dataset presenta problemas de calidad de datos que deberán resolverse antes de modelar (especialmente la corrección de escalas y el tratamiento de nulos).
- Existen amplias referencias y experiencias previas con este dataset que servirán de guía.
- Las métricas RMSE y R² son adecuadas para evaluar el desempeño del modelo.

## 💡 Proposals and Ideas

A partir de los resultados obtenidos, se propone lo siguiente para las siguientes etapas del proyecto:

1. **Corrección de escalas** (Tarea 2 - Exploración inicial): normalizar las columnas que presentan valores en escala ×1000. Verificar cada columna contra los rangos documentados del dataset original.

2. **Estrategia de imputación** (Tarea 2): evaluar si los valores nulos en `medv` se imputan o si se eliminan esos registros, considerando que son el 3.6% del total.

3. **Análisis de outliers** (Tarea 3 - EDA): realizar un análisis detallado con boxplots y tests estadísticos para decidir el tratamiento de valores atípicos, en particular para `crim`.

4. **Modelo heurístico baseline** (Tarea 5): implementar un modelo simple basado en las reglas manuales descritas anteriormente (precio base + ajustes por `rm`, `lstat`, `crim`) para tener una referencia antes del modelado ML.

5. **Variables clave a considerar**: con base en la literatura, priorizar `rm`, `lstat` y `ptratio` como predictores principales en el feature engineering.

6. **Consideración ética**: documentar y discutir el uso de la variable `black` y evaluar si su inclusión es apropiada para el modelo.

## 📖 References

- Harrison, D. and Rubinfeld, D.L. (1978). *Hedonic housing prices and the demand for clean air*. Journal of Environmental Economics and Management, 5, 81-102.
- Scikit-learn documentation: [sklearn.datasets.load_boston (deprecated)](https://scikit-learn.org/1.1/modules/generated/sklearn.datasets.load_boston.html)
- UCI Machine Learning Repository: [Housing Data Set](https://archive.ics.uci.edu/ml/machine-learning-databases/housing/)
- Guía del proyecto del curso: [joserzapata.github.io/post/ciencia-datos-proyecto-python/1-data/](https://joserzapata.github.io/post/ciencia-datos-proyecto-python/1-data/)
- Gitflow branching model: [joserzapata.github.io/courses/ciencia-datos-en-produccion/control-versiones/branching-model/](https://joserzapata.github.io/courses/ciencia-datos-en-produccion/control-versiones/branching-model/)